Update kubectl debug docs for 1.20 release (#24847)
kubernetes/website: content/en/docs/tasks/debug-application-cluster/debug-running-pod.md

Adding a new container can be useful when your application is running but not
behaving as you expect and you'd like to add additional troubleshooting
utilities to the Pod.

For example, maybe your application's container images is built on busybox
but you need debugging utilities not included in busybox. You can simulate
this scenario using `kubectl run`:

```shell
kubectl run myapp --image=busybox --restart=Never -- sleep 1d
```

Run this command to create a copy of `myapp` named `myapp-copy` that adds a
new ubuntu container for debugging.

```shell
kubectl debug myapp -it --image=ubuntu --share-processes --copy-to=myapp-debug
```

```
Defaulting debug container name to debugger-w7xmf.
If you don't see a command prompt, try pressing enter.
root@myapp-debug:/#
```

There are a few things worth noting in this invocation of `kubectl debug`:

* `kubectl debug` automatically generates a container name  if you don't choose
  one using the `--container` flag.
* The `-i` flag causes `kubectl debug` to attach to the new container by
  default.  You can prevent this by specifying `--attach=false`. If your session
  becomes disconnected you can reattach using `kubectl attach`.

> * The `--share-processes` causes the Pod to use one process namespace so that

Reviewer: nit: Line 210, extra space before `if`
nit: The `--share-processes` flag
nit: ... causes the Pod to use one process per namespace. You can identify other processes from other
containers using `ps`.

Author: Fixed the first two, but the third suggestion is incorrect. I've updated the text to:

> The `--share-processes` flag causes the Pod to use a single process namespace. You can identify other processes from other containers using `ps`.

I'm not sure if this addresses your initial concern, though.

Reviewer: OK. I am not sure what exactly is a "single process namespace"?
Single process + one namespace or ?

Author: Under the hood it sets the `shareProcessNamespace` option in [v1.PodSpec](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#podspec-v1-core). A "process namespace" is how the kernel keeps application processes in containers separate from processes in other containers and the larger system.

By default every container gets a separate "process namespace" so that there's no visibility to processes that are running in different containers. That makes debugging across container boundaries more difficult.

When one uses `--share-processes`, the kubelet will use one "process namespace" for all of the containers in the Pod so that processes from all of the containers in that Pod are visible to each other.

Reviewer: Great. Thanks for the explanation!

Author: @kbhawkey Do you think I should rephrase/expand on this concept in the text or is it good as is?

Reviewer: I'll read through the page ...

To simulate a crashing application, use `kubectl run` to create a container
that immediately exists:

```
kubectl run --image=busybox myapp -- false
```

You can see using `kubectl describe pod myapp` that this container is crashing:

```
Containers:
  myapp:
    Image:         busybox
    ...
    Args:
      false
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
```

You can use `kubectl debug` to create a copy of this Pod with the command
changed to an interactive shell:

```
kubectl debug myapp -it --copy-to=myapp-debug --container=myapp -- sh
```

```
If you don't see a command prompt, try pressing enter.
/ #
```

Now you have an interactive shell that you can use to perform tasks like
checking filesystem paths or running the container command manually. There
are a few things worth noting about this invocation of `kubectl debug`:

* Since you want to change the command of a specific container you must
  specify its name using `--container` or `kubectl debug` will instead
  create a new container to run the command you specified.

> * The `-i` flag causes `kubectl debug` to attach to the container by default.  You can prevent this by specifying `--attach=false`. If your session becomes

Reviewer: Line 271, Could you add a carriage return before, "You can prevent ..."

Author: Wow. Yes. Not sure how that happened.

  disconnected you can reattach using `kubectl attach`.

Don't forget to clean up the debugging Pod when you're finished with it:

```shell
kubectl delete pod myapp myapp-debug
```

### Copying a Pod while changing container images

In some situations you may want to change a misbehaving Pod from its normal
production container images to an image containing a debugging build or
additional utilities.

As an example, create a Pod using `kubectl run`:

```
kubectl run myapp --image=busybox --restart=Never -- sleep 1d
```

Now use `kubectl debug` to make a copy and change its container image

> to ubuntu:

Reviewer: nit: Either use Ubuntu or add code formatting for the value as `ubuntu`.

Author: Fixed.

```
kubectl debug myapp --copy-to=myapp-debug --set-image=*=ubuntu
```

The syntax of `--set-image` uses the same `container_name=image` syntax as
`kubectl set image`. `*=ubuntu` means change the image of all containers

> to `ubuntu.

Reviewer: nit: missing a back tick?

Author: Added, thanks.

Don't forget to clean up the debugging Pod when you're finished with it:

```shell
kubectl delete pod myapp myapp-debug
```

## Debugging via a shell on the node {#node-shell-session}

If none of these approaches work, you can find the Node one which the Pod is
running and create a privileged Pod running in the host namespaces. To create
an interactive shell on a node using `kubectl debug`, run:

```shell
kubectl debug node/mynode -it --image=ubuntu
```

```
Creating debugging pod node-debugger-mynode-pdx84 with container debugger on node mynode.
If you don't see a command prompt, try pressing enter.
root@ek8s:/#
```

When creating a debugging session on a node, keep in mind that:

* `kubectl debug` will automatically generate the name of the new Pod based on
  the name of the Node.

> * The container will run in the host IPC, Network & PID namespaces.

Reviewer: nit: s/&/and/

Author: Fixed.

Don't forget to clean up the debugging Pod when you're finished with it:

```shell
kubectl delete pod myapp myapp-debug
```

### Copying a Pod while changing its command

Sometimes it's useful to change the command for a container, for example to
add a debugging flag or because the application is crashing.

To simulate a crashing application, use `kubectl run` to create a container
that immediately exists:

```
kubectl run --image=busybox myapp -- false
```

You can see using `kubectl describe pod myapp` that this container is crashing:

```
Containers:
  myapp:
    Image:         busybox
    ...
    Args:
      false
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
```

You can use `kubectl debug` to create a copy of this Pod with the command
changed to an interactive shell:

```
kubectl debug myapp -it --copy-to=myapp-debug --container=myapp -- sh
```

```
If you don't see a command prompt, try pressing enter.
/ #
```

Now you have an interactive shell that you can use to perform tasks like
checking filesystem paths or running the container command manually. There

> are a few things worth noting about this invocation of `kubectl debug`:

Reviewer 2: Make this a note.

```suggestion
{{< note >}}
```

> * Since you want to change the command of a specific container you must

Reviewer 2: ```suggestion
* To change the command of a specific container you must
```

  specify its name using `--container` or `kubectl debug` will instead
  create a new container to run the command you specified.
* The `-i` flag causes `kubectl debug` to attach to the container by default.
  You can prevent this by specifying `--attach=false`. If your session becomes

>   disconnected you can reattach using `kubectl attach`.

Reviewer 2: This section is well constructed. Nice work! ✨

Author: Thanks! I'm slowly getting a feel for website style 😀

> 

Reviewer 2: ```suggestion

{{< /note >}}

```

Don't forget to clean up the debugging Pod when you're finished with it:

```shell
kubectl delete pod myapp myapp-debug
```

### Copying a Pod while changing container images

In some situations you may want to change a misbehaving Pod from its normal
production container images to an image containing a debugging build or
additional utilities.

As an example, create a Pod using `kubectl run`:

```
kubectl run myapp --image=busybox --restart=Never -- sleep 1d
```

Now use `kubectl debug` to make a copy and change its container image
to `ubuntu`:

```
kubectl debug myapp --copy-to=myapp-debug --set-image=*=ubuntu
```

The syntax of `--set-image` uses the same `container_name=image` syntax as
`kubectl set image`. `*=ubuntu` means change the image of all containers
to `ubuntu`.

Don't forget to clean up the debugging Pod when you're finished with it:

```shell
kubectl delete pod myapp myapp-debug
```

## Debugging via a shell on the node {#node-shell-session}

> If none of these approaches work, you can find the Node one which the Pod is

Reviewer 2: Grammar nit
```suggestion
If none of these approaches work, you can find the Node on which the Pod is
```

running and create a privileged Pod running in the host namespaces. To create
an interactive shell on a node using `kubectl debug`, run:

```shell
kubectl debug node/mynode -it --image=ubuntu
```

```
Creating debugging pod node-debugger-mynode-pdx84 with container debugger on node mynode.
If you don't see a command prompt, try pressing enter.
root@ek8s:/#
```

When creating a debugging session on a node, keep in mind that:

> * `kubectl debug` will automatically generate the name of the new Pod based on

Reviewer 2: Avoid future tense
```suggestion
* `kubectl debug` automatically generates the name of the new Pod based on
```

  the name of the Node.

> * The container will run in the host IPC, Network, and PID namespaces.

Reviewer 2: Avoid future tense
```suggestion
* The container runs in the host IPC, Network, and PID namespaces.
```

```

```
Defaulting debug container name to debugger-w7xmf.
If you don't see a command prompt, try pressing enter.
root@myapp-debug:/#
```

{{< note >}}
* `kubectl debug` automatically generates a container name if you don't choose
  one using the `--container` flag.
* The `-i` flag causes `kubectl debug` to attach to the new container by
  default.  You can prevent this by specifying `--attach=false`. If your session
  becomes disconnected you can reattach using `kubectl attach`.
* The `--share-processes` allows the containers in this Pod to see processes
  from the other containers in the Pod. For more information about how this
  works, see [Share Process Namespace between Containers in a Pod](
  /docs/tasks/configure-pod-container/share-process-namespace/).
{{< /note >}}

Don't forget to clean up the debugging Pod when you're finished with it:

```shell
kubectl delete pod myapp myapp-debug
```

### Copying a Pod while changing its command

Sometimes it's useful to change the command for a container, for example to
add a debugging flag or because the application is crashing.

To simulate a crashing application, use `kubectl run` to create a container

> that immediately exists:

Reviewer 3: There is a typo here, it should **exit's** rather than **exists**

Author: @shubhamrajvanshi Thanks for pointing this out. I opened #25458 to fix.

kubectl delete pod ephemeral-demo
```

## Debugging using a copy of the Pod

Sometimes Pod configuration options make it difficult to troubleshoot in certain

> situations. For example, `kubectl exec` won't work with a container image that

Reviewer: nit: [what situations?]: Sometimes Pod configuration options make it difficult to troubleshoot containers and applications running in containers?
nit: For example, you can't run `kubectl exec` to troubleshoot your container if your container image
does not include a shell or if your application crashes on startup.

Author: Applied the second suggestion, but I don't understand the first one. Is that the syntax for a reference link?

Reviewer: Sorry @verb, I did not get a chance to look at the changes again before it was approved 😞

Reviewer 4: Eh sorry... 😭

doesn't include a shell or if an application crashes on startup. In these
situations you can use `kubectl debug` to create a copy of the Pod with
configuration values changed to aid debugging.

### Copying a Pod while adding a new container

Adding a new container can be useful when your application is running but not
behaving as you expect and you'd like to add additional troubleshooting
utilities to the Pod.

> For example, maybe your application's container images are built on busybox

Reviewer: nit: format `busybox`?

Author: Fixed

but you need debugging utilities not included in busybox. You can simulate
this scenario using `kubectl run`:

```shell
kubectl run myapp --image=busybox --restart=Never -- sleep 1d
```

Run this command to create a copy of `myapp` named `myapp-copy` that adds a
new Ubuntu container for debugging:

```shell
kubectl debug myapp -it --image=ubuntu --share-processes --copy-to=myapp-debug
```

```
Defaulting debug container name to debugger-w7xmf.
If you don't see a command prompt, try pressing enter.
root@myapp-debug:/#
```

{{< note >}}

> 

Reviewer: nit: You can remove the line before and after the note text.

Reviewer: nit:  You may want to add a single sentence to introduce the list of flag descriptions.

It may be helpful to provide more context about the `share-processes` flag.
What do you think about:

The `--share-processes` flag causes the Pod to share a single process namespace between all containers in a Pod.
A single process namespace lets a container view and signal processes from other containers in the same Pod.
You can identify processes from other containers using `ps`.
For more information about `shareProcessNamespace`, see the [PodSpec](/docs/reference/generated/kubernetes-api/{{< param "version" >}}/#podspec-v1-core) API.

From the API reference:
```
Share a single process namespace between all of the containers in a pod. When this is set containers will be able to view and signal processes from other containers in the same pod,
```

Author: * Removed the extra space
* The single sentence to introduce the list of flag descriptions was what another reviewer suggested I replace with the `note` tags. 😃  I'll leave this for now.
* For `--share-processes` I've replaced the description of the flag with a short reason why you'd want to use it here, and then linked to the `ShareProcessNamespace` docs for full context. I like the way it turned out, PTAL.
